File: iOS/Views/HistoryViewModel.swift
import Foundation
import Observation
import SwiftData

/// Dữ liệu điểm trên biểu đồ.
public struct ChartData: Identifiable, Sendable {
    public let date: Date
    public let count: Int
    public var id: Date { date }
}

public struct InsightsTrendPoint: Identifiable, Sendable {
    public let date: Date
    public let takenCount: Int
    public let skippedCount: Int
    public var id: Date { date }
}

public struct InsightsItem: Identifiable, Sendable {
    public let title: String
    public let count: Int
    public var id: String { title }
}

public struct InsightsSummary: Sendable {
    public let windowDays: Int
    public let takenCount: Int
    public let skippedCount: Int
    public let lateCount: Int
    public let completionRate: Double
    public let topSkipped: [InsightsItem]
    public let topLate: [InsightsItem]
    public let topLateHour: InsightsItem?
}

/// ViewModel xử lý logic lịch sử uống trên iOS.
@Observable
@MainActor
public final class HistoryViewModel {
    // MARK: - State
    public var weeklyData: [ChartData] = []
    public var insights7: InsightsSummary?
    public var insights30: InsightsSummary?
    public var trend7: [InsightsTrendPoint] = []
    public var trend30: [InsightsTrendPoint] = []
    
    public init() {}
    
    // MARK: - Logic
    
    /// Tổng hợp dữ liệu uống theo tuần.
    public func processHistory(records: [IntakeRecord]) {
        weeklyData = weeklyChartData(records: records)
        insights7 = buildInsights(records: records, windowDays: 7)
        insights30 = buildInsights(records: records, windowDays: 30)
        trend7 = trendData(records: records, windowDays: 7)
        trend30 = trendData(records: records, windowDays: 30)
    }

    private func weeklyChartData(records: [IntakeRecord]) -> [ChartData] {
        let calendar = Calendar.current
        let todayStart = calendar.startOfDay(for: .now)
        let start = calendar.date(byAdding: .day, value: -6, to: todayStart) ?? todayStart
        let isDescending = (records.first?.date ?? .distantPast) >= (records.last?.date ?? .distantPast)
        var counts: [Date: Int] = [:]
        counts.reserveCapacity(7)
        for record in records {
            let day = calendar.startOfDay(for: record.date)
            if day < start {
                if isDescending { break }
                continue
            }
            counts[day, default: 0] += 1
        }
        return (0..<7).reversed().compactMap { offset in
            guard let date = calendar.date(byAdding: .day, value: -offset, to: todayStart) else { return nil }
            return ChartData(date: date, count: counts[date] ?? 0)
        }
    }

    private func buildInsights(records: [IntakeRecord], windowDays: Int) -> InsightsSummary? {
        guard windowDays > 0 else { return nil }
        let calendar = Calendar.current
        let todayStart = calendar.startOfDay(for: .now)
        guard let start = calendar.date(byAdding: .day, value: -(windowDays - 1), to: todayStart) else { return nil }
        let isDescending = (records.first?.date ?? .distantPast) >= (records.last?.date ?? .distantPast)
        var taken = 0
        var skipped = 0
        var late = 0
        var skippedCounts: [String: Int] = [:]
        var lateCounts: [String: Int] = [:]
        var lateHourCounts: [Int: Int] = [:]
        for record in records {
            let day = calendar.startOfDay(for: record.date)
            if day > todayStart { continue }
            if day < start {
                if isDescending { break }
                continue
            }
            let name = record.supplement?.name ?? "not_available".localized
            if record.status == IntakeStatus.taken.rawValue {
                taken += 1
                if isLateTaken(record) {
                    late += 1
                    lateCounts[name, default: 0] += 1
                    let updatedAt = Date(timeIntervalSince1970: TimeInterval(record.updatedAtEpochMs) / 1000)
                    let hour = calendar.component(.hour, from: updatedAt)
                    lateHourCounts[hour, default: 0] += 1
                }
                continue
            }
            if record.status == IntakeStatus.skipped.rawValue {
                skipped += 1
                skippedCounts[name, default: 0] += 1
            }
        }
        if taken == 0, skipped == 0 { return nil }
        guard let completionRate = DoseTimingPolicy.completionRate(taken: taken, skipped: skipped) else { return nil }
        return InsightsSummary(
            windowDays: windowDays,
            takenCount: taken,
            skippedCount: skipped,
            lateCount: late,
            completionRate: completionRate,
            topSkipped: topItemsFromCounts(skippedCounts),
            topLate: topItemsFromCounts(lateCounts),
            topLateHour: topLateHourFromCounts(lateHourCounts)
        )
    }

    private func isLateTaken(_ record: IntakeRecord) -> Bool {
        DoseTimingPolicy.isLateTaken(
            status: record.status,
            scheduledAtEpochMs: Int64(record.date.timeIntervalSince1970 * 1_000),
            updatedAtEpochMs: record.updatedAtEpochMs
        )
    }

    private func topItemsFromCounts(_ counts: [String: Int]) -> [InsightsItem] {
        counts
            .map { InsightsItem(title: $0.key, count: $0.value) }
            .sorted {
                guard $0.count != $1.count else { return $0.title < $1.title }
                return $0.count > $1.count
            }
            .prefix(3)
            .map { $0 }
    }

    private func topLateHourFromCounts(_ counts: [Int: Int]) -> InsightsItem? {
        guard let best = counts.max(by: { $0.value < $1.value }) else { return nil }
        let label = String(format: "%02d:00", best.key)
        return InsightsItem(title: label, count: best.value)
    }

    private func trendData(records: [IntakeRecord], windowDays: Int) -> [InsightsTrendPoint] {
        guard windowDays > 0 else { return [] }
        let calendar = Calendar.current
        let todayStart = calendar.startOfDay(for: .now)
        guard let start = calendar.date(byAdding: .day, value: -(windowDays - 1), to: todayStart) else { return [] }
        let isDescending = (records.first?.date ?? .distantPast) >= (records.last?.date ?? .distantPast)
        var counts: [Date: (taken: Int, skipped: Int)] = [:]
        counts.reserveCapacity(windowDays)
        for record in records {
            let day = calendar.startOfDay(for: record.date)
            if day > todayStart { continue }
            if day < start {
                if isDescending { break }
                continue
            }
            if record.status == IntakeStatus.taken.rawValue {
                let current = counts[day] ?? (0, 0)
                counts[day] = (current.taken + 1, current.skipped)
                continue
            }
            if record.status == IntakeStatus.skipped.rawValue {
                let current = counts[day] ?? (0, 0)
                counts[day] = (current.taken, current.skipped + 1)
            }
        }
        return (0..<windowDays).reversed().compactMap { offset in
            guard let date = calendar.date(byAdding: .day, value: -offset, to: todayStart) else { return nil }
            let value = counts[date] ?? (0, 0)
            return InsightsTrendPoint(date: date, takenCount: value.taken, skippedCount: value.skipped)
        }
    }
}

File: Android/app/src/main/java/com/example/supplementtracker/presentation/home/HistoryViewModel.kt
package com.example.supplementtracker.presentation.home

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.supplementtracker.domain.repository.IntakeRecord
import com.example.supplementtracker.domain.repository.SupplementRepository
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.flowOf
import kotlinx.coroutines.flow.flatMapLatest
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.mapLatest
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.time.LocalDate
import java.time.ZoneId
import com.example.supplementtracker.presentation.navigation.ActiveClientManager
import com.example.supplementtracker.service.CoachClientSnapshot
import com.example.supplementtracker.service.CoachOverviewBuilder
import com.example.supplementtracker.service.CoachOverviewSummary
import com.example.supplementtracker.service.CoachWorkspaceSource
import com.example.supplementtracker.service.CoachWorkspaceSourceProvider
import com.example.supplementtracker.service.RepositoryCoachWorkspaceSourceProvider
import com.example.supplementtracker.service.CommercialFeature
import com.example.supplementtracker.service.CommercialPlan
import com.example.supplementtracker.service.EntitlementManager
import com.example.supplementtracker.service.EntitlementPolicy
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.first
import java.time.format.TextStyle
import java.util.Locale

/**
 * Dữ liệu cho biểu đồ lịch sử.
 */
data class HistoryChartData(
    val label: String,
    val count: Int
)

/**
 * Trạng thái UI cho màn hình lịch sử.
 */
sealed class HistoryUiState {
    data object Loading : HistoryUiState()
    data object NoClient : HistoryUiState()
    data class Success(
        val chartData: List<HistoryChartData>,
        val sections: List<HistorySection>,
        val insights7: InsightsSummary?,
        val insights30: InsightsSummary?,
        val trend7: List<InsightsTrendPoint>,
        val trend30: List<InsightsTrendPoint>,
        val analyticsAvailable: Boolean
    ) : HistoryUiState()
}

data class HistorySection(
    val date: LocalDate,
    val records: List<IntakeRecord>
)

data class InsightsTrendPoint(
    val date: LocalDate,
    val takenCount: Int,
    val skippedCount: Int
)

data class InsightsItem(
    val title: String,
    val count: Int
)

data class InsightsSummary(
    val windowDays: Int,
    val completionRate: Float,
    val takenCount: Int,
    val skippedCount: Int,
    val lateCount: Int,
    val topSkipped: List<InsightsItem>,
    val topLate: List<InsightsItem>,
    val topLateHour: InsightsItem?
)

sealed class CoachOverviewUiState {
    data object Idle : CoachOverviewUiState()
    data object Loading : CoachOverviewUiState()
    data object Locked : CoachOverviewUiState()
    data object Error : CoachOverviewUiState()
    data class Ready(val summary: CoachOverviewSummary) : CoachOverviewUiState()
}

/**
 * ViewModel xử lý lịch sử uống cho Android.
 */
@OptIn(ExperimentalCoroutinesApi::class)
class HistoryViewModel(
    private val repository: SupplementRepository,
    private val activeClientManager: ActiveClientManager,
    private val entitlementManager: EntitlementManager,
    private val coachSourceProvider: CoachWorkspaceSourceProvider = RepositoryCoachWorkspaceSourceProvider(repository)
) : ViewModel() {
    private val mutableCoachOverview = MutableStateFlow<CoachOverviewUiState>(CoachOverviewUiState.Idle)
    val coachOverview: StateFlow<CoachOverviewUiState> = mutableCoachOverview.asStateFlow()
    private val mutableCoachWindowDays = MutableStateFlow(7)
    val coachWindowDays: StateFlow<Int> = mutableCoachWindowDays.asStateFlow()
    private var coachSource: CoachWorkspaceSource? = null

    val uiState: StateFlow<HistoryUiState> = combine(
        activeClientManager.currentClientId,
        entitlementManager.snapshot
    ) { clientId, entitlement -> clientId to entitlement.plan }
        .flatMapLatest { (clientId, plan) ->
            val id = clientId?.toString() ?: return@flatMapLatest flowOf(HistoryUiState.NoClient)
            repository.observeAllRecordsByClient(id)
                .mapLatest { records -> withContext(Dispatchers.Default) { processHistory(records, plan) } }
        }
        .stateIn(
            scope = viewModelScope,
            started = SharingStarted.WhileSubscribed(5000),
            initialValue = HistoryUiState.Loading
        )

    fun refreshCoachOverview() {
        if (!entitlementManager.canUse(CommercialFeature.COACH_REPORTS)) {
            mutableCoachOverview.value = CoachOverviewUiState.Locked
            return
        }
        viewModelScope.launch {
            mutableCoachOverview.value = CoachOverviewUiState.Loading
            try {
                coachSource = coachSourceProvider.load()
                mutableCoachOverview.value = CoachOverviewUiState.Ready(buildCoachOverview())
            } catch (error: CancellationException) {
                throw error
            } catch (_: Exception) {
                mutableCoachOverview.value = CoachOverviewUiState.Error
            }
        }
    }

    fun selectCoachWindow(windowDays: Int) {
        if (windowDays !in setOf(7, 30, 90) || mutableCoachWindowDays.value == windowDays) return
        mutableCoachWindowDays.value = windowDays
        val source = coachSource ?: return
        mutableCoachOverview.value = CoachOverviewUiState.Ready(buildCoachOverview(source))
    }

    private fun buildCoachOverview(
        source: CoachWorkspaceSource = coachSource ?: CoachWorkspaceSource(emptyList(), emptyMap())
    ): CoachOverviewSummary {
        return CoachOverviewBuilder.build(
            clients = source.clients,
            recordsByClient = source.records,
            nowEpochMs = System.currentTimeMillis(),
            windowDays = mutableCoachWindowDays.value
        )
    }

    fun coachClientDetail(clientId: java.util.UUID): com.example.supplementtracker.service.CoachClientDetail? {
        val source = coachSource ?: return null
        val client = source.clients.firstOrNull { it.id == clientId } ?: return null
        return com.example.supplementtracker.service.CoachWorkspaceBuilder.buildDetail(
            client = client,
            records = source.records[clientId].orEmpty(),
            nowEpochMs = System.currentTimeMillis(),
            windowDays = mutableCoachWindowDays.value
        )
    }

    private fun processHistory(records: List<IntakeRecord>, plan: CommercialPlan): HistoryUiState {
        val zoneId = ZoneId.systemDefault()
        val today = LocalDate.now(zoneId)
        val historyStart = today.minusDays(EntitlementPolicy.historyDays(plan) - 1)
        val orderedRecords = records.sortedByDescending { it.date }.filter { record ->
            val day = java.time.Instant.ofEpochMilli(record.date).atZone(zoneId).toLocalDate()
            !day.isBefore(historyStart) && !day.isAfter(today)
        }
        val startDate = today.minusDays(6)
        val counts = mutableMapOf<LocalDate, Int>()
        val sections = mutableListOf<HistorySection>()
        var currentDate: LocalDate? = null
        var currentBucket = mutableListOf<IntakeRecord>()
        for (record in orderedRecords) {
            val date = java.time.Instant.ofEpochMilli(record.date).atZone(zoneId).toLocalDate()
            if (!date.isBefore(startDate) && !date.isAfter(today)) {
                counts[date] = (counts[date] ?: 0) + 1
            }
            if (currentDate == null) {
                currentDate = date
            }
            if (date != currentDate) {
                val finishedDate = currentDate ?: date
                sections.add(HistorySection(date = finishedDate, records = currentBucket.toList()))
                currentDate = date
                currentBucket = mutableListOf()
            }
            currentBucket.add(record)
        }
        val lastDate = currentDate
        if (lastDate != null) {
            sections.add(HistorySection(date = lastDate, records = currentBucket.toList()))
        }

        val chartData = mutableListOf<HistoryChartData>()
        for (i in 6 downTo 0) {
            val date = today.minusDays(i.toLong())
            chartData.add(HistoryChartData(label = dayLabel(date), count = counts[date] ?: 0))
        }

        val analyticsAvailable = EntitlementPolicy.allows(plan, CommercialFeature.ADHERENCE_ANALYTICS)
        val insights7 = if (analyticsAvailable) buildInsights(orderedRecords, 7, zoneId) else null
        val insights30 = if (analyticsAvailable) buildInsights(orderedRecords, 30, zoneId) else null
        val trend7 = if (analyticsAvailable) buildTrend(orderedRecords, 7, zoneId) else emptyList()
        val trend30 = if (analyticsAvailable) buildTrend(orderedRecords, 30, zoneId) else emptyList()
        return HistoryUiState.Success(
            chartData, sections, insights7, insights30, trend7, trend30, analyticsAvailable
        )
    }

    private fun buildInsights(
        records: List<IntakeRecord>,
        windowDays: Long,
        zoneId: ZoneId
    ): InsightsSummary? {
        val today = LocalDate.now(zoneId)
        val start = today.minusDays(windowDays - 1)
        val window = records.filter {
            val day = java.time.Instant.ofEpochMilli(it.date).atZone(zoneId).toLocalDate()
            !day.isBefore(start) && !day.isAfter(today)
        }
        if (window.isEmpty()) return null
        val taken = window.count { it.status == "Taken" }
        val skipped = window.count { it.status == "Skipped" }
        val late = window.count { isLateTaken(it) }
        val denom = (taken + skipped).coerceAtLeast(1)
        val completion = taken.toFloat() / denom.toFloat()
        return InsightsSummary(
            windowDays = windowDays.toInt(),
            completionRate = completion,
            takenCount = taken,
            skippedCount = skipped,
            lateCount = late,
            topSkipped = topListBySupplement(window, "Skipped", limit = 3),
            topLate = topLateBySupplement(window, limit = 3),
            topLateHour = topLateHour(window, zoneId)
        )
    }

    private fun buildTrend(
        records: List<IntakeRecord>,
        windowDays: Long,
        zoneId: ZoneId
    ): List<InsightsTrendPoint> {
        val today = LocalDate.now(zoneId)
        val start = today.minusDays(windowDays - 1)
        val counts = mutableMapOf<LocalDate, Pair<Int, Int>>()
        for (record in records) {
            val day = java.time.Instant.ofEpochMilli(record.date).atZone(zoneId).toLocalDate()
            if (day.isAfter(today)) continue
            if (day.isBefore(start)) break
            val current = counts[day] ?: (0 to 0)
            if (record.status == "Taken") {
                counts[day] = (current.first + 1) to current.second
            } else if (record.status == "Skipped") {
                counts[day] = current.first to (current.second + 1)
            }
        }
        val points = mutableListOf<InsightsTrendPoint>()
        for (i in (windowDays - 1) downTo 0) {
            val day = today.minusDays(i)
            val value = counts[day] ?: (0 to 0)
            points.add(InsightsTrendPoint(date = day, takenCount = value.first, skippedCount = value.second))
        }
        return points
    }

    private fun isLateTaken(record: IntakeRecord): Boolean {
        if (record.status != "Taken") return false
        if (record.updatedAtEpochMs <= 0L) return false
        val threshold = record.date + 20 * 60 * 1000
        return record.updatedAtEpochMs > threshold
    }

    private fun topListBySupplement(records: List<IntakeRecord>, status: String, limit: Int): List<InsightsItem> {
        if (limit <= 0) return emptyList()
        val grouped = records.filter { it.status == status }
            .groupBy { it.supplementName ?: "N/A" }
            .mapValues { it.value.size }
            .entries
            .sortedByDescending { it.value }
            .take(limit)
        return grouped.map { InsightsItem(title = it.key, count = it.value) }
    }

    private fun topLateHour(records: List<IntakeRecord>, zoneId: ZoneId): InsightsItem? {
        val late = records.filter { isLateTaken(it) }
        if (late.isEmpty()) return null
        val grouped = late.groupBy {
            val hour = java.time.Instant.ofEpochMilli(it.updatedAtEpochMs).atZone(zoneId).hour
            "%02d:00".format(hour)
        }.mapValues { it.value.size }
        val best = grouped.maxByOrNull { it.value } ?: return null
        return InsightsItem(title = best.key, count = best.value)
    }

    private fun topLateBySupplement(records: List<IntakeRecord>, limit: Int): List<InsightsItem> {
        if (limit <= 0) return emptyList()
        val late = records.filter { isLateTaken(it) }
        if (late.isEmpty()) return emptyList()
        val grouped = late.groupBy { it.supplementName ?: "N/A" }
            .mapValues { it.value.size }
            .entries
            .sortedByDescending { it.value }
            .take(limit)
        return grouped.map { InsightsItem(title = it.key, count = it.value) }
    }

    private fun dayLabel(date: LocalDate): String {
        val locale = Locale.getDefault()
        return date.dayOfWeek.getDisplayName(TextStyle.SHORT, locale)
    }
}
